sys/net/application_layer/nanocoap/nanocoap.c

```diff
@@ -545,29 +545,73 @@ ssize_t coap_tree_handler(coap_pkt_t *pkt, uint8_t *resp_buf, unsigned resp_buf_
     return coap_build_reply(pkt, COAP_CODE_404, resp_buf, resp_buf_len, 0);
 }
 
+ssize_t coap_build_reply_header(coap_pkt_t *pkt, unsigned code,
+                                void *buf, size_t len,
+                                int ct,
+                                void **payload, size_t *payload_len_max)
+{
+    uint8_t *bufpos = buf;
+    uint32_t no_response;
+    uint8_t type = coap_get_type(pkt) == COAP_TYPE_CON
+                 ? COAP_TYPE_ACK
+                 : COAP_TYPE_NON;
+
+    bufpos += coap_build_hdr(buf, type, coap_get_token(pkt), coap_get_token_len(pkt),
```

Comment:
From the `coap_build_hdr` docs:

> Caller must ensure `hdr` can hold the header and the full token!

You also don't check whether `buf` is large enough to hold the content type (format) option. `*payload_len_max` might underflow.

Reply:
> You also don't check whether buf is large enough to hold the content type (format) option.

Our CoAP API is unfortunately kind of insane in that it expects you to write to a buffer and then check *after* the fact that it overflowed.
There is no way to detect if the result of `coap_put_option_ct()` will fit into the supplied buffer beforehand.

Reply:
> Our CoAP API is unfortunately kind of insane in that it expects you to write to a buffer and then check after the fact that it overflowed.

Oh wow, that's crazy o.O Any plans for this to be fixed on our roadmap?

> There is no way to detect if the result of `coap_put_option_ct()` will fit into the supplied buffer beforehand.

According to [the RFC](https://datatracker.ietf.org/doc/html/rfc7252#section-5.10) it can at most be 2B long if I read it correctly.